src/crawlee/storages/_key_value_store.py

```diff
@@ -171,3 +178,39 @@ async def get_public_url(self, key: str) -> str:
             The public URL for the given key.
         """
         return await self._resource_client.get_public_url(key)
+
+    async def get_auto_saved_value(self, key: str, default_value: dict[str, Any] | None = None) -> dict[str, Any]:
```

Comment:
Currently, crawlee-js does this in two separate ways. What you ported over is the "regular" one. Apart from that, we have adaptive crawling, where `useState` bypasses `get_auto_saved_value`. See https://github.com/apify/crawlee/blob/master/packages/playwright-crawler/src/internals/adaptive-playwright-crawler.ts for details. 

The general idea behind this is that we don't want to "save" the changed state before we know that the request handler succeeded.

We should probably discuss what would be the best way forward. Try to give the javascript adaptive crawler a glance, but if it doesn't make any sense, just say so so that we don't waste any time. Then we'll talk more :slightly_smiling_face:

Reply:
I looked up adaptive-playwright-crawler, if I understand its workings correctly, it completely overrides the default `useState`

We could do the same thing when we need to implement this in Python. Given that this behavior is specific to one particular implementation of `crawler`.